### PR TITLE
Start tests with react

### PR DESCRIPTION
**Pull request by garmin-mjames**
Currently tests are started with `add`. This causes problems with how
certain Triggers behave, namely `Event` and `NullTrigger`.

They behave differently at the beginning of a test than after the test
awaits a `GPITrigger`. Using `react` to start the test makes the
triggers behave the same during the entirety of the test.

Trigger behavior is different when in a `react` event loop versus a straight `schedule` called through `add`.

Found while discussing scheduler details in #1541. Thanks to @ktbarrett for `NullTrigger` test.

### CODE REVIEW

File: cocotb/scheduler.py [hunk 1]
```diff
@@ -547,12 +547,18 @@ def _coroutine_yielded(self, coro, trigger):
...
+
+    def _trigger_task_with_outcome(self, task, trigger_name=None, outcome=None):
+        """Trigger task with new outcome"""
+        new_trigger = NullTrigger(name=trigger_name, outcome=outcome)
+
+        self._coro2trigger[task] = new_trigger
+        self._trigger2coros[new_trigger] = [task]
+
+        # wake up the Task
+        self.react(new_trigger)
```
Comment (eric-wieser) [inline cocotb/scheduler.py:560]: I'd prefer we write this in terms of `prime`, else we end up being woken by an unprimed trigger.

**Comment by garmin-mjames** [comment]
There's some issue with `test_closedown`.

**Comment by ktbarrett** [comment]
I'm not near my computer to try, but couldn't you just use `queue` and `react`, so you dont have to add another method?

**Comment by ktbarrett** [comment]
@garmin-mjames I've tracked down the issue in `test_closedown` to the implementation of `_finished_scheduler`. I'm not sure how it even worked beforehand, that deserves some investigation. Just add `cocotb.regression_manager.handle_result(self._test)` after the test abortion and it will pass.

**Comment by ktbarrett** [comment]
Perhaps this is a good time to address the issue in `test_closedown`. This point of it is to have the TB close the simulator and stop cocotb gracefully, but it contains a bunch of additional "stuff" in the test that has somehow been obscuring and implementation bug in `_finish_test`. Perhaps this test should be simplified to just enter and then wait for a long time.

**Comment by eric-wieser** [comment]
I'd probably err on the side of duplicating the test and making the copy simpler, just in case the current form contains an additional corner case we forget about.

File: cocotb/scheduler.py [hunk 2]
```diff
@@ -547,12 +547,18 @@ def _coroutine_yielded(self, coro, trigger):
...
+
+    def _trigger_task_with_outcome(self, task, trigger_name=None, outcome=None):
+        """Trigger task with new outcome"""
+        new_trigger = NullTrigger(name=trigger_name, outcome=outcome)
+
+        self._coro2trigger[task] = new_trigger
+        self._trigger2coros[new_trigger] = [task]
+
+        # wake up the Task
+        new_trigger.prime(self.react)
```
Comment (eric-wieser) [inline cocotb/scheduler.py:560]: Why not just:
```suggestion
        self._coroutine_yielded(task, new_trigger)
```
optimization doesn't seem particularly worthwhile here.

Comment (eric-wieser) [reply cocotb/scheduler.py:560]: Perhaps `_coroutine_yielded(coro, trigger)` should be renamed something like `_resume_coro_upon(coro, trigger)`.

Comment (ktbarrett) [reply cocotb/scheduler.py:560]: If he does that he could just inline the method as well.

**Comment by ktbarrett** [comment]
I think the other elements there are to ensure that forked coroutines and externals are killed correctly when the test is ended via this method. If we only had one entry point for test end it would be redundant.

**Comment by garmin-mjames** [comment]
Why isn't `Clock.start()` being `fork`ed in `test_closedown`?

**Comment by ktbarrett** [comment]
@garmin-mjames A leftover from when clocks were implemented in C and didn't need to be forked.

**Comment by garmin-mjames** [comment]
This in `_sim_event` doesn't appear to be the case, at least with icarus:
https://github.com/cocotb/cocotb/blob/beb42627bdc6d6d0271b51f2903a885706eeb5cd/cocotb/__init__.py#L198-L199

File: tests/test_cases/test_closedown/test_closedown.py [hunk 3]
```diff
@@ -57,12 +51,20 @@ def clock_mon(dut):
...
-    clock = Clock(dut.clk, 100)
-    clock.start()
+    """Allow the dut to call system tasks from verilog.
+    $fatal() will fail the test, and scheduler will cleanup forked coroutines.
+    """
+    cocotb.fork(Clock(dut.clk, 100).start())
     cocotb.fork(clock_mon(dut))
     cocotb.fork(run_external(dut))
-    yield Timer(10000000)
+    yield Timer(10000000, units='ns')
```
Comment (ktbarrett) [inline tests/test_cases/test_closedown/test_closedown.py:63]: Why don't you make this `async`/`await` now. It's on our Python 2 cleanup todo list.

File: tests/test_cases/test_closedown/test_closedown.py [hunk 4]
```diff
@@ -57,12 +51,20 @@ def clock_mon(dut):
...
+
+
+@cocotb.test()
+async def test_after_system_task_fail(dut):
+    """Test to run after failed test.
+    """
```
Comment (ktbarrett) [inline tests/test_cases/test_closedown/test_closedown.py:69]:
```suggestion
    """
    Test to run after failed test.
    """
```

File: documentation/source/newsfragments/1705.bugfix.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+Scheduling behavior in the beginning of a test, before the first ``await`` of a :class:`~cocotb.triggers.GPITrigger`, now matches the rest of the test.
```
Comment (ktbarrett) [inline documentation/source/newsfragments/1705.bugfix.rst:1]:
```suggestion
Scheduling behavior is now consistent before and after the first ``await`` of a :class:`~cocotb.triggers.GPITrigger`.
```
?

File: tests/test_cases/test_cocotb/test_scheduler.py [hunk 6]
```diff
@@ -206,3 +206,59 @@ async def killer():
...
+    Test that Event.set() doesn't cause an immediate reschedule.
+
+    The coroutine waiting with Event.wait() will be woken after
+    the current coroutine awaits a trigger.
+    """
+
+    e = Event()
+    event_set = False
+
+    @cocotb.coroutine
```
Comment (ktbarrett) [inline tests/test_cases/test_cocotb/test_scheduler.py:249]: Don't need this anymore.

File: tests/test_cases/test_cocotb/test_scheduler.py [hunk 7]
```diff
@@ -206,3 +206,59 @@ async def killer():
...
+async def test_event_set_schedule(dut):
+    """
+    Test that Event.set() doesn't cause an immediate reschedule.
+
+    The coroutine waiting with Event.wait() will be woken after
+    the current coroutine awaits a trigger.
+    """
+
+    e = Event()
+    event_set = False
```
Comment (ktbarrett) [inline tests/test_cases/test_cocotb/test_scheduler.py:247]: `Event` has a `.fired` attribute, use that instead.

**Comment by ktbarrett** [comment]
Good catch using `check_termination`, there is a bunch of cleanup that needs to be done between tests. Is there a way to move this to use `react` as well? I haven't looked at it enough. Ideally we will only have one external entry into the scheduler, so there is only one place where we need to check for test completion. That can be done in another PR if you feel this is good for now.

File: tests/test_cases/test_cocotb/test_scheduler.py (hunk 7 above)
Comment (ktbarrett) [reply tests/test_cases/test_cocotb/test_scheduler.py:247]: It should probably also have a `is_set()` method to mirror asyncio.

File: documentation/source/newsfragments/1705.bugfix.rst (hunk 5 above)
Comment (garmin-mjames) [reply documentation/source/newsfragments/1705.bugfix.rst:1]: Should we add bugfix info about `SIM_TEST_FAIL` too?

**Comment by garmin-mjames** [comment]
There are several paths between the Scheduler and RegressionManager, and having `add` break into the scheduler complicates it further. I think after `add`/`fork` is changed and further decoupling happens things will shake out and we can simplify further.

File: tests/test_cases/test_cocotb/test_scheduler.py (hunk 7 above)
Comment (garmin-mjames) [reply tests/test_cases/test_cocotb/test_scheduler.py:247]: I actually want to check if the other coro/task wakes up. I'll fix the naming.

**Comment by garmin-mjames** [comment]
I don't think that we can use `react` without changing how aborting a test works. Currently it just unschedules the test, so the scheduler has do deal with the cleanup afterwords.

There have been discussions in the past about raising an exception in the test that can be used for cleanup etc. That would match what the scheduler does for a Trigger.prime() failure or if the test awaits an invalid object. Then cleanup after the test exits (and raise another Exception if it tries to ignore the abort and await another trigger?).

File: documentation/source/newsfragments/1705.bugfix.rst (hunk 5 above)
Comment (ktbarrett) [reply documentation/source/newsfragments/1705.bugfix.rst]: That's a good idea, I'm not sure how to have multiple newsfrags per PR in the same category though.

**Comment by ktbarrett** [comment]
@garmin-mjames If we add a `_check_termination` or some similar "check if running test is complete" method to `react`, simply `abort`ing the current test will work. But let's leave that for another PR.

File: documentation/source/newsfragments/1705.bugfix.rst (hunk 5 above)
Comment (garmin-mjames) [reply documentation/source/newsfragments/1705.bugfix.rst]: I could break the `finish_test` changes into another PR and get that in first, or we can just use the combined newsfragment here.

Comment (ktbarrett) [reply documentation/source/newsfragments/1705.bugfix.rst]: [This](https://cocotb--1705.org.readthedocs.build/en/1705/release_notes.html#bugfixes) is what is currently looks like, we don't have 2 bullet points. I'll that up to @eric-wieser.

File: tests/test_cases/test_cocotb/test_scheduler.py [hunk 8]
```diff
@@ -206,3 +206,60 @@ async def killer():
...
+        await event.wait()
+        nonlocal waiter_scheduled
+        waiter_scheduled = True
+
+    cocotb.fork(waiter(e))
+
+    e.set()
+
+    # waiter() shouldn't run until after test awaits a trigger
+    assert waiter_scheduled == False
```
Comment (ktbarrett) [inline tests/test_cases/test_cocotb/test_scheduler.py:261]:
```suggestion
    assert waiter_scheduled is False
```

File: tests/test_cases/test_cocotb/test_scheduler.py [hunk 9]
```diff
@@ -206,3 +206,60 @@ async def killer():
...
+
+    await NullTrigger()
+
+    assert waiter_scheduled == True
```
Comment (ktbarrett) [inline tests/test_cases/test_cocotb/test_scheduler.py:265]:
```suggestion
    assert waiter_scheduled is True
```

File: cocotb/scheduler.py [hunk 10]
```diff
@@ -519,7 +519,7 @@ def _schedule_write(self, handle, write_func, *args):
         self._write_calls.append((write_func, args))
         self._writes_pending.set()
 
-    def _coroutine_yielded(self, coro, trigger):
+    def _resume_coro_upon(self, coro, trigger):
         """Prime the trigger and update our internal mappings."""
```
Comment (eric-wieser) [inline cocotb/scheduler.py:523]:
```suggestion
        """Schedule `coro` to be resumed when `trigger` fires. """
```

File: documentation/source/newsfragments/1705.bugfix.rst (hunk 5 above)
Comment (eric-wieser) [reply documentation/source/newsfragments/1705.bugfix.rst]: I think you can create a second fragment, `1705.bugfix.1.rst`

**Comment by eric-wieser** [comment]
I don't fully understand the `check_termination()` change, and haven't followed all the discussion here. If you haven't already posted it, can you post the logs of `test_closedown` failing before that change?

**Comment by garmin-mjames** [comment]
Log of `test_closedown` using `react` to start test:
[sim.log](https://github.com/cocotb/cocotb/files/4553331/sim.log)
Log of `test_closedown` after applying the cleanup commit but not the `_check_termination` call:
[cleanup_sim.log](https://github.com/cocotb/cocotb/files/4553332/cleanup_sim.log)

What's happening:

1. `$fatal()` calls up to `finish_test`, which calls `abort` on the test
2. `abort` calls `unschedule` which calls `cleanup`
3. The result isn't handled by `RegressionManager`, so it returns to the simulator
4. Simulator has no callbacks queued up due to `cleanup`, so it ends simulation and calls `finish_scheduler`

`_check_termination` catches the fact that the test was aborted and continues the test running loop back into the `RegressionManager`, allowing further tests to run.

**Comment by eric-wieser** [comment]
Release notes seem to work: https://cocotb--1705.org.readthedocs.build/en/1705/release_notes.html#bugfixes

Will come back and review those logs  in the next day or so.

**Comment by eric-wieser** [comment]
I wonder if this fixes what's left with #637

**Comment by garmin-mjames** [comment]
I wonder how many `yield Timer(1)`s at the beginning of tests are because of this

**Comment by eric-wieser** [comment]
That, is the right :)

File: documentation/source/newsfragments/1705.bugfix.1.rst [hunk 11]
```diff
@@ -0,0 +1 @@
+Scheduler now correctly cleans up when a test is failed using the ``$fatal()`` System Task in Verilog.
```
Comment (cmarqu) [inline documentation/source/newsfragments/1705.bugfix.1.rst:1]:
```suggestion
The scheduler now correctly cleans up when a test is failed using the ``$fatal()`` System Task in Verilog.
```

Comment (eric-wieser) [reply documentation/source/newsfragments/1705.bugfix.1.rst:1]: Is this actually a bug that was present in 1.3? Or just a bug that was introduced by the commit before it?

Comment (eric-wieser) [reply documentation/source/newsfragments/1705.bugfix.1.rst:1]: The latter should not be in a release note.

Comment (garmin-mjames) [reply documentation/source/newsfragments/1705.bugfix.1.rst:1]: @eric-wieser I guess that's true, it worked before. I'll remove the newsfragment.

**Comment by garmin-mjames** [comment]
Rebased on master.

**Comment by eric-wieser** [comment]
Thanks for tracking down what I could not!

Would appreciate if you could follow up with a PR to remove all the gh-637 hacks (should be able to search for `gh-637` or `#637`)
